server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,105 @@
+import com.google.common.jimfs.Jimfs;
+import io.deephaven.plugin.type.JsPlugin;
+import io.deephaven.plugin.type.JsPluginRegistration;
+import org.eclipse.jetty.servlet.DefaultServlet;
+import org.eclipse.jetty.servlet.ServletHolder;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link JsPluginRegistration} for Jetty.
+ *
+ * @see #servletHolder(String)
+ */
+public final class JsPlugins implements JsPluginRegistration {
+
+    /**
+     * Creates a new js plugins instance with an in-memory filesystem.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final FileSystem fs = Jimfs.newFileSystem(Configuration.unix());
+        final JsPlugins jsPlugins = new JsPlugins(fs.getPath("/"));
+        jsPlugins.writeManifest(); // empty manifest
+        return jsPlugins;
+    }
+
+    private final Path path;
+    private final Map<String, JsPlugin> plugins;
+
+    private JsPlugins(Path path) {
+        this.path = Objects.requireNonNull(path);
+        this.plugins = new LinkedHashMap<>();
+    }
+
+    /**
+     * Creates a {@link DefaultServlet} servlet holder.
+     *
+     * @param name the name
+     * @return the servlet holder
+     */
+    public ServletHolder servletHolder(String name) {
+        final ServletHolder jsPlugins = new ServletHolder(name, DefaultServlet.class);
+        jsPlugins.setInitParameter("resourceBase", path.toUri().toString());
+        jsPlugins.setInitParameter("pathInfoOnly", "true");
+        jsPlugins.setInitParameter("dirAllowed", "false");
+        jsPlugins.setAsyncSupported(true);
+        return jsPlugins;
+    }
+
+    @Override
+    public synchronized void register(JsPlugin jsPlugin) {
+        if (plugins.containsKey(jsPlugin.name())) {
+            throw new IllegalArgumentException(
+                    String.format("js plugin with name '%s' already exists", jsPlugin.name()));
+        }
+        try {
+            jsPlugin.copyTo(path.resolve(jsPlugin.name()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        plugins.put(jsPlugin.name(), jsPlugin);
+        try {
+            writeManifest();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeManifest() throws IOException {
+        // note: would this be better as a custom servlet instead? (this current way is certainly easy...)
+        try (final OutputStream out = new BufferedOutputStream(Files.newOutputStream(path.resolve("manifest.json")))) {
```

Comment:
buffering to an in-memory fs?

i ask because its going to be fully read into memory either way, and afaict never GC'd - and while the other files aren't going to be very small, this one is likely to be no bigger than a few kb?

Reply:
I'm taking the FS-agnostic approach here - mentioned earlier, but this was originally written for local temporary filesystem. Even though it's not technically necessary when writing to an in-memory fs, I think it's still best practice to buffer like this.

Reply:
Makes sense, I added notes in a few other places where we should be truly agnostic to follow this attitude.